src/common/parser.c

```diff
@@ -0,0 +1,399 @@
+		return PARSER_SIZE_PATH_EXPECTED; /* no path */
+
+	LOG(10, "size '%s' path '%s'", size_str, path_str);
+
+	/*
+	 * A format of the size is checked in detail. As regards the path,
+	 * it is checked only if the read path is an absolute path.
+	 * The rest should be checked during creating/opening the file.
+	 */
+
+	/* check if the read path is an absolute path */
+	if (path_str[0] != '/')
+		return PARSER_WRONG_PATH; /* must be an absolute path */
+
+	*path = Strdup(path_str);
+
+	/* check format of size */
+	*size = strtoull(size_str, &endptr, 10);
+	if (endptr && endptr[0] != '\0') {
+		if (endptr[1] != '\0' || strchr("KMGT", endptr[0]) == NULL)
+			return PARSER_WRONG_SIZE; /* wrong format of size */
+		/* multiply size by a unit */
+		switch (endptr[0]) {
+			case 'K':
+				*size *= 1ULL << 10; /* 1 KB */
+				break;
+			case 'M':
+				*size *= 1ULL << 20; /* 1 MB */
+				break;
+			case 'G':
+				*size *= 1ULL << 30; /* 1 GB */
+				break;
+			case 'T':
+				*size *= 1ULL << 40; /* 1 TB */
+				break;
+		}
+	}
+
+	return PARSER_CONTINUE;
+}
+
+/*
+ * parser_add_part -- (internal) add a new part to the list
+ */
+static void
+parser_add_part(struct headparts *headp, int id, size_t size, char *path)
+{
+	struct part *p = Malloc(sizeof (*p));
+	if (p == NULL) {
+		FATAL("!Malloc");
+	}
+	p->id = id;
+	p->size = size;
+	p->path = path;
+	TAILQ_INSERT_TAIL(headp, p, next);
+}
+
+/*
+ * parser_add_replica -- (internal) add a new replica to the list
+ */
+static struct replica *
+parser_add_replica(struct headreps *headr, int id)
+{
+	struct replica *r = Malloc(sizeof (*r));
+	if (r == NULL) {
+		FATAL("!Malloc");
+	}
+	r->rep_id = id;
+	r->nparts = 0;
+	r->rep_size = 0;
+	TAILQ_INIT(&r->rep_parts);
+	TAILQ_INSERT_TAIL(headr, r, next);
+	return TAILQ_LAST(headr, headreps);
+}
+
+/*
+ * parser_init_poolset -- init a pool set structure
+ */
+void
+parser_init_poolset(struct poolset *ps)
+{
+	TAILQ_INIT(&ps->set_parts);
+	TAILQ_INIT(&ps->reps);
+	ps->nparts = 0;
+	ps->set_size = 0;
+	ps->nreps = 0;
+}
+
+/*
+ * parser_free_poolset -- free a pool set structure
+ */
+void
+parser_free_poolset(struct poolset *ps)
+{
+	struct part *p;
+	struct replica *r;
+
+	/* free pool set parts */
+	while (!TAILQ_EMPTY(&ps->set_parts)) {
+		p = TAILQ_FIRST(&ps->set_parts);
+		TAILQ_REMOVE(&ps->set_parts, p, next);
+		Free(p->path);
+		Free(p);
+	}
+
+	/* free all replicas */
+	while (!TAILQ_EMPTY(&ps->reps)) {
+		r = TAILQ_FIRST(&ps->reps);
+		/* free parts of the current replica */
+		while (!TAILQ_EMPTY(&r->rep_parts)) {
+			p = TAILQ_FIRST(&r->rep_parts);
+			TAILQ_REMOVE(&r->rep_parts, p, next);
+			Free(p->path);
+			Free(p);
+		}
+		TAILQ_REMOVE(&ps->reps, r, next);
+		Free(r);
+	}
+}
+
+/*
+ * parser_parse_set_file -- parse a pool set file
+ */
+int
+parser_parse_set_file(const char *path, int fd, struct poolset *ps)
+{
+	LOG(4, "parsing file %s", path);
+
+#define	REPLICA_HDR_SIG "REPLICA"
+#define	REPLICA_HDR_SIG_LEN 7
+#define	PARSER_MAX_LINE 1024
```

Comment:
`1024` is probably too short, since `PATH_MAX` is usually `4096` and a line could have a path name and a size separated by whitespace.